### PR TITLE
Multicast reliability, association, and message order guarantee changes.  Change LargeSample test to check for in-order messages.

### PR DESCRIPTION
**Pull request by oschwaldp-oci**

### CODE REVIEW

File: dds/DCPS/transport/framework/TransportClient.cpp [hunk 1]
```diff
@@ -496,7 +496,8 @@ TransportClient::PendingAssoc::initiate_connect(TransportClient* tc,
           data_.remote_reliable_, data_.remote_durable_};
 
         TransportImpl::AcceptConnectResult res;
-
+        GuidConverter tmp_local(tc->repo_id_);
+        GuidConverter tmp_remote(this->data_.remote_id_);
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/TransportClient.cpp]: Why were these moved?

File: dds/DCPS/transport/framework/TransportSendBuffer.cpp [hunk 2]
```diff
@@ -295,6 +295,10 @@ SingleSendBuffer::resend(const SequenceRange& range, DisjointSequence* gaps)
 bool
 SingleSendBuffer::resend_i(const SequenceRange& range, DisjointSequence* gaps)
 {
+  //Special case, nak to make sure it has all history
+  SequenceNumber lowForAllResent;
+  range.first == SequenceNumber() ? lowForAllResent = low() : lowForAllResent = range.first;
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/TransportSendBuffer.cpp]: This seems overly complex... how about
const SequenceNumber lowForAllResent = range.first == SequenceNumber() ? low() : range.first;

File: dds/DCPS/transport/multicast/MulticastDataLink.cpp [hunk 3]
```diff
@@ -296,10 +297,10 @@ MulticastDataLink::sample_received(ReceivedDataSample& sample)
 
       for (MulticastSessionMap::iterator it(temp_sessions.begin());
           it != temp_sessions.end(); ++it) {
-        it->second->control_received(sample.header_.submessage_id_,
-            sample.sample_);
+        bool ctrl_rcvd = it->second->control_received(sample.header_.submessage_id_,
+                                                      sample.sample_);
         // reset read pointer
-        if (ptr) {
+        if (ctrl_rcvd && ptr) {
```
Comment (mitza-oci) [inline dds/DCPS/transport/multicast/MulticastDataLink.cpp]: What's the rationale behind not resetting rd_ptr when control_received returns false?  Maybe this deserves clarification in the comment on ln 302 (or I'm missing something obvious).

File: dds/DCPS/transport/multicast/MulticastDataLink.cpp [hunk 4]
```diff
@@ -315,10 +333,89 @@ MulticastDataLink::sample_received(ReceivedDataSample& sample)
...
+      this->session_lock_, false);
+  SequenceNumber seq = data.header_.sequence_;
+  MulticastSessionMap::iterator session_it = sessions_.find(source);
+  if (session_it != sessions_.end()) {
+    bool reliable = session_it->second->is_reliable();
+    if (!reliable) {
+      return true;
+    }
+    MulticastSession_rch session(session_it->second);
+    DisjointSequence recvd = this->data_samples_seen_[data.header_.publication_id_];
```
Comment (mitza-oci) [inline dds/DCPS/transport/multicast/MulticastDataLink.cpp]: Is a copy of the DisjointSequence needed here?  That could be expensive (heap alloc).

File: dds/DCPS/transport/framework/TransportClient.cpp (hunk 1 above)
Comment (oschwaldp-oci) [inline dds/DCPS/transport/framework/TransportClient.cpp]: tc is a simple TransportClient*, and during the connect_datalink call (inside tc->initiate_connect_i) the guard is released at which time the TransportClient could go away and the connect fails.  This path is for failure to connect/initiate_connect_i path and thus should not access the tc pointer after this call if AcceptConnectResult returned was false.  This allows the stack to unwind without accessing anything inside tc.

File: dds/DCPS/transport/framework/TransportSendBuffer.cpp (hunk 2 above)
Comment (oschwaldp-oci) [inline dds/DCPS/transport/framework/TransportSendBuffer.cpp]: That's more concise.  Thanks.

File: dds/DCPS/transport/multicast/MulticastDataLink.cpp (hunk 4 above)
Comment (oschwaldp-oci) [inline dds/DCPS/transport/multicast/MulticastDataLink.cpp]: Indeed. Thanks for the catch.
